core/sql/optimizer/NATable.cpp

```diff
@@ -3553,6 +3553,48 @@ NAType* getSQColTypeForHive(const char* hiveType, NAMemory* heap)
   if ( !strcmp(hiveType, "timestamp"))
     return new (heap) SQLTimestamp(TRUE /* allow NULL */ , 6, heap);
 
+  if ( !strcmp(hiveType, "date"))
+    return new (heap) SQLDate(TRUE /* allow NULL */ , heap);
+
+  if ( !strncmp(hiveType, "varchar", 7) )
+  {
+    char maxLen[32];
+    memset(maxLen, 0, 32);
+    int i=0,j=0;
+    int copyit = 0;
+
+    //get length
+    for(i = 0; i < strlen(hiveType) ; i++)
+    {
+      if(hiveType[i] == '(') //start
+      {
+        copyit=1;
+        continue;
+      }
+      else if(hiveType[i] == ')') //stop
+        break; 
+      if(copyit > 0)
+      {
+        maxLen[j] = hiveType[i];
+        j++;
+      }
+    }
+    Int32 len = atoi(maxLen);
+
+    if(len == 0) return NULL;  //cannot parse correctly
+
+    NAString hiveCharset =
+        ActiveSchemaDB()->getDefaults().getValue(HIVE_DEFAULT_CHARSET);
+
+    return new (heap) SQLVarChar(CharLenInfo((hiveCharset == CharInfo::UTF8 ? 0 : len),len),
```

Comment:
One solution is to do this mapping:

Hive STRING ==> Trafodion VARCHAR(n BYTES), with n determined by HIVE_MAX_STRING_LENGTH

Hive VARCHAR(n) ==> Trafodion VARCHAR(n)

This treats STRING and VARCHAR differently, the max size of STRING is in bytes, the max size of VARCHAR is in characters. This is not ideal, users will probably expect string + CQD HIVE_MAX_STRING_LENGTH 'n' to work similar to VARCHAR(n)?

There is no good way to achieve this, though, at least not without wasting space and making incompatible changes.

I wonder what other people think.